### DISCUSSION TITLE
Forme with an associated object & nested attributes

### DISCUSSION THREAD

**Discussion by teamschlaflos** [jeremyevans/forme#123] [General]
Trying to get Frome to work with a model class which has a `many_to_one` association.  I'm using `forme_set` to update all direct model attributes, which works fine.

When adding a field to the form, from an associated model & using a nested_attribute, the select box renders fine and lists options, but storing the user input returns an error:

```
NoMethodError: undefined method `pk' for "064fcc28-c006-40ff-bcc8-e0c17f09b6ce":String (NoMethodError)
raise(Error, "associated object #{o.inspect} does not have a primary key") if o && !o.pk
n/vendor/ruby/3.2.0/gems/sequel-5.80.0/lib/sequel/model/associations.rb:2878:in `set_associated_object'
n/vendor/ruby/3.2.0/gems/sequel-5.80.0/lib/sequel/model/associations.rb:2338:in `block in def_many_to_one'
n/vendor/ruby/3.2.0/gems/sequel-5.80.0/lib/sequel/model/base.rb:1573:in `block in set_fields'
n/vendor/ruby/3.2.0/gems/sequel-5.80.0/lib/sequel/model/base.rb:1573:in `each'
n/vendor/ruby/3.2.0/gems/sequel-5.80.0/lib/sequel/model/base.rb:1573:in `set_fields'
n/vendor/ruby/3.2.0/gems/forme-2.3.0/lib/roda/plugins/forme_set.rb:63:in `forme_set'
n/routes/admin.rb:160:in `block (5 levels) in <class:N>'
```

The documentation states that this is expected behavior:

> Additionally, forme_set does not currently handle subform/nested_attributes.
>
> In cases where forme_set does not handle things correctly, you can use forme_parse, which will return metadata that forme_set uses (forme_set calls forme_parse internally). forme_parse returns a hash with the following keys:
>
> `:values` - A hash of values that can be used to update the model, suitable for passing to Sequel::Model#set.
> `:validations` - A hash of values suitable for merging into forme_validations. Used to check that the submitted values for associated objects match one of the options for the input in the form.
>
> It is possible to use forme_set for the values it can handle, and set other fields manually using set_fields.

However I couldn't find an example, and I'm unable to figure out how one would use forme_set & set_fields (I'm usually using the specs for digging up workable code).

Basically I'd need to create one form, some fields should use forme_set, some fields should be ignored and only be processed by a manual set_fields.

**Comment by jeremyevans** [comment]
I think the last paragraph in the documentation you quoted is not worded correctly.  It should be changed to:

```
It is possible to use forme_set for the forms it can handle, and use forme_parse and
set_fields for other forms.
```

I apologize for the confusion and will push a fix shortly.

As the first line in the documentation you quoted states, you cannot use `forme_set`, as it doesn't handled `nested_attributes`, you need to use `forme_parse`. To combine `forme_parse` and `set_fields`, you might be able to do something like:

```ruby
values, validations = obj.forme_parse(hash).values_at(:values, :validations)
obj.set_fields(values, %w'field1 field2 ...')
obj.forme_validations.merge!(validations)
```

**Reply by jeremyevans** [reply]
Fixed by 0ae6563c412fb5827dd981681fb5f37f9449c5e1

**Comment by teamschlaflos** [comment]
The documentation is a bit sparse about this topic.

I found that even your previous example didn't work, as I believe I should rather use the Roda "frome_set" plugin version of [`forme_parse`](https://github.com/jeremyevans/forme/blob/master/lib/roda/plugins/forme_set.rb#L55C1-L56C1) instead of the sequel plugin [`forme_parse`](https://github.com/jeremyevans/forme/blob/master/lib/sequel/plugins/forme_set.rb#L63).

So the following code works for updating the direct name attribute & allow to set (or unset) the associated Artist object.

```ruby
class Album < Sequel::Model
  many_to_one :artist

  plugin :forme
  plugin :forme_set
  plugin :nested_attributes

  nested_attributes :artist

  def validate
    super
    errors.add(:name, "validation should always add an error")
  end
end
```

```ruby
r.post do
  @album = Album[r.params["id"]]

  # Parse _forme_set_data from r.params, verify csrf & hmac then return the values
  values, validations = forme_parse(@album).values_at(:values, :validations)

  # Convert keys back to string as roda/plugins/forme_set transforms keys to symbols
  # set_fields requires strings as keys
  values.transform_keys!(&:to_s) 

  # Set the album.name field based on the form input
  @album.set_fields(values, %w'name')

  # Update album.artist association, use `active_support presence` here to avoid
  # @album.artist = r.params["album"]["artist"].empty" ? nil : Artist[r.params["album"]["artist]
  @album.artist = Artist[r.params["album"]["artist"].presence]

  # TODO: does not work
  @album.forme_validations.merge!(validations)
  @album.artist.validate

  @album.artist.save
end
```

```ruby
<%=
  form(@album) do |f|
    f.input(:name, key: "name", type: :text)
    f.input(:artist, key: "artist")
  end
%>
```

I'm still not able to get the validation to work (rendering the form doesn't add any error elements). 

What I noticed is that upon form submission the payload will contain `_forme_set_data.valid_values` only for the artist, and so does the validations variable from the code above.

**Reply by jeremyevans** [reply]
In this example, you are only setting values in the album's table (name and artist_id).  If you remove the `key` options in the form, I would think that `forme_set` would work correctly.  Your example doesn't use nested_attributes in the form.  I'm not sure why you are validating/saving the artist, since you aren't making modifications to the artist, you are only setting the `artist_id` foreign key for the album.
